hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java

```diff
@@ -0,0 +1,171 @@
+            checkForSlowOperations(partitionOperationThreads, partitionCurrentOperationData);
+
+            if (nowNanos - lastLogPurge > logPurgeIntervalNanos) {
+                lastLogPurge = nowNanos;
+                purgeLogs();
+            }
+
+            try {
+                TimeUnit.NANOSECONDS.sleep(ONE_SECOND_IN_NANOS - (System.nanoTime() - nowNanos));
+            } catch (InterruptedException ignored) {
+                break;
+            }
+        }
+    }
+
+    Collection<SlowOperationLog> getSlowOperationLogs() {
+        return slowOperationLogs.values();
+    }
+
+    public void shutdown() {
+        running = false;
+    }
+
+    private void initCurrentOperationData(CurrentOperationData[] currentOperationDataArray) {
+        for (int i = 0; i < currentOperationDataArray.length; i++) {
+            currentOperationDataArray[i] = new CurrentOperationData();
+            currentOperationDataArray[i].operationHashCode = -1;
+        }
+    }
+
+    private void checkForSlowOperations(BasicOperationScheduler.OperationThread[] operationThreads,
+                                        CurrentOperationData[] currentOperationDataArray) {
+        for (int i = 0; i < operationThreads.length; i++) {
+            Object currentOperation = operationThreads[i].currentOperation;
+            if (currentOperation == null) {
+                continue;
+            }
+
+            CurrentOperationData currentOperationData = currentOperationDataArray[i];
+            int operationHashCode = System.identityHashCode(currentOperation);
+            if (currentOperationData.operationHashCode != operationHashCode) {
+                // initialize currentOperationData
+                currentOperationData.operationHashCode = operationHashCode;
+                currentOperationData.startNanos = nowNanos;
+                currentOperationData.invocation = null;
+                continue;
+            }
+
+            long duration = nowNanos - currentOperationData.startNanos;
+            if (duration < slowOperationThresholdNanos) {
+                continue;
+            }
+
+            SlowOperationLog.Invocation invocation = currentOperationData.invocation;
+            if (invocation != null) {
+                // update existing invocation to avoid creation of stack trace
+                invocation.duration = TimeUnit.NANOSECONDS.toMillis(duration);
+                invocation.lastAccess = nowNanos;
+                continue;
+            }
+
+            SlowOperationLog log = getOrCreateSlowOperationLog(operationThreads[i], currentOperation);
+            currentOperationData.invocation = log.getOrCreateInvocation(operationHashCode, duration, nowNanos, nowMillis);
+
+            logger.warning("Slow operation detected!\n" + log);
+        }
+    }
+
+    private SlowOperationLog getOrCreateSlowOperationLog(Thread operationThread, final Object operation) {
+        StringBuilder stackTrace = new StringBuilder();
```

Comment:
I need to think about this class because are going to create quite a lot of litter. We can reuse the StringBuilder btw; since only a single thread will be using it.